File: Project_Update_2/cppClusterProto/cluster_proto.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <cmath>
#include <vector>

#define MAX_POINTS 360
#define MAX_CLUSTERS 20

typedef struct {
    int num_points;
    std::vector<int> points;
} cluster_t;

void clusterOp(const std::vector<double>& angles, const std::vector<double>& distances, std::vector<cluster_t>& clusters) {
    const double D_th = 200.0;
    const int minGroupSize = 4;

    int num_points = angles.size();
    std::vector<bool> grouped(num_points, false);

    // Process LIDAR data into clusters
    for (int point = 0; point < num_points; point++) {
        // If point has already been grouped, move onto the next point
        if (grouped[point]) {
            continue;
        }

        cluster_t cluster;
        cluster.num_points = 0;

        // Initialize the points vector of the cluster_t struct
        cluster.points.clear();
        //cluster.points.push_back(point);

        // Test in counter clockwise direction
        int test_point = point;
        for (int i = 0; i < num_points; i++) {
            if (grouped[i]) {
                continue;
            }

            double dTheta = angles[i] - angles[test_point];
            double d1 = distances[test_point];
            double d2 = distances[i];
            double d3 = sqrt(d1*d1 + d2*d2 - 2*d1*d2*cos(dTheta));

            if (d3 < D_th && !grouped[i]) {
                std::cout << i << " ";
                cluster.num_points++;
                cluster.points.push_back(i);
                grouped[i] = true;
                test_point = i;
            }
        }

        if (cluster.num_points >= minGroupSize) {
            clusters.push_back(cluster);
            if (clusters.size() == MAX_CLUSTERS && num_points > MAX_CLUSTERS) {
                break;
            }
        }
    }
}


int main() {
    std::vector<double> angle;
    std::vector<double> distance;

    // Open the CSV file for reading
    std::ifstream infile("data.csv");

    // Read each line of the file
    std::string line;
    while (std::getline(infile, line)) {

        // Parse the line into two integers using a stringstream
        std::stringstream ss(line);
        std::string value;
        std::getline(ss, value, ',');
        int a = std::stoi(value);
        std::getline(ss, value, ',');
        int d = std::stoi(value);

        // Add the integers to the angle and distance vectors
        angle.push_back(a);
        distance.push_back(d);
    }

    /*
    // Print the angle and distance vectors for testing
    std::cout << "Angle: ";
    for (int i = 0; i < angle.size(); i++) {
        std::cout << angle[i] << " ";
    }
    std::cout << std::endl;

    std::cout << "Distance: ";
    for (int i = 0; i < distance.size(); i++) {
        std::cout << distance[i] << " ";
    }
    std::cout << std::endl;
    */

    std::vector<cluster_t> clusters;
    clusterOp(angle, distance, clusters);
    

    for (int i = 0; i < clusters.size(); i++) {
        std::cout << "Cluster " << i << ": ";
        for (int j = 0; j < clusters[i].points.size(); j++) {
            std::cout << clusters[i].points[j] << " ";
        }
        std::cout << std::endl;
    }

    return 0;

}

File: Project_Update_1/clusterProto.py
import PyLidar3
import matplotlib.pyplot as plt
import numpy as np
import math
import time
import csv
import os
import shutil

# ls /dev/tty*
# /dev/tty.usbserial-0001


def clusterOp(angle, distance, D_th = 100, minGroupSize = 4):
    objects = []
    grouped = set()

    for point in range(0,360):

        #If point has already been grouped move onto the next point
        if point in grouped:
            continue

        cluster = set()
        cluster.add(point)

        #Test in counter clockwise direction
        for i in range(0,360):
            #print(point)
            dTheta = angle[i] - angle[point]
            d1 = distance[point]
            d2 = distance[i]
            d3 = math.sqrt(d1**2 + d2**2 - 2*d1*d2*math.cos(dTheta))

            
            if d3 < D_th:
                cluster.add(i)
                grouped.add(i)
                point = i

        if len(cluster) >= minGroupSize:
            objects.append(cluster)

    return objects

def plotObjects(objects, meanObjects = "True"):
    meanTheta = []
    meanDistance = []

    for i in range(len(objects)):
        cluster = objects[i]
        clusterTheta = []
        clusterDistance = []

        for j in range(len(cluster)):
            cluster = list(cluster)
            i = cluster[j]
            clusterTheta.append(angle[i])
            clusterDistance.append(distance[i])
        ax.set_theta_zero_location("S")
        ax.scatter(clusterTheta, clusterDistance, s = 5)

        if meanObjects:
            meanTheta.append(np.mean(clusterTheta))
            meanDistance.append(np.mean(clusterDistance))
            ax.scatter(meanTheta, meanDistance, s = 40, marker="^")

def recordData(directoryName, distance):
    try:
        if not os.path.isdir('LidarData'):
            print(0)
            os.makedirs('LidarData')
        
        if not os.path.isdir('./LidarData/'+directoryName):
            os.makedirs('./LidarData/'+directoryName)

        sampleNum = 0
        while os.path.exists('./LidarData/'+directoryName+'/'+ directoryName + '%s.csv' % str(sampleNum).zfill(3)):
            sampleNum += 1

        
        with open('./LidarData/'+directoryName+'/'+ directoryName + '%s.csv' % str(sampleNum).zfill(3), 'w', newline='') as file:
        
            writer = csv.writer(file)

            for point in range(len(distance)):
                writer.writerow([point, distance[point]])




    except OSError:
        pass

def clearData(directoryName):
    try:
        
        if os.path.isdir('./LidarData/'+directoryName):
            shutil.rmtree('./LidarData/'+directoryName)
    except OSError:
        pass

print('Store Data?')
storeData = input('Confirm (1)/ Cancel (0): ')
if storeData == "1":
    sampleName = input('Enter name: ')
    clearData(sampleName)


x=[]
y=[]
for _ in range(360):
    x.append(0)
    y.append(0)

angle = []
for i in range(0,360):
    angle.append(math.radians(i))




#port = input("Enter port name which lidar is connected:") 
port = "/dev/tty.usbserial-0001"
Obj = PyLidar3.YdLidarX4(port, 12000) 
if(Obj.Connect()):


    fig = plt.figure()
    ax = fig.add_subplot(projection='polar')
    


    print(Obj.GetDeviceInfo())
    gen = Obj.StartScanning()
    t = time.time() # start time 
    while (time.time() - t) < 30: #scan for 30 seconds
        #print(next(gen))
        data = next(gen)
        distance = []
        for points in data:
            distance.append(data[points])

        distance = np.flip(distance)
        objectsCurrent = clusterOp(angle,distance)

        if storeData == "1":
            recordData(sampleName, distance)

        plotObjects(objectsCurrent, True)
        # ax.scatter(angle,distance,c='r',s=5)
        plt.pause(0.1)
        time.sleep(1)
        ax.cla()  

        

    
    Obj.StopScanning()
    Obj.Disconnect()    
else:
    print("Error connecting to device")
